### PR TITLE
Update retry policy for transient errors and store to table storage

### PR DESCRIPTION
**Pull request by abdullah248**
## Description
Update the retry policy for transient errors to use exponential bakeoff (currently retries after 2, 4 and 8 seconds but can be adjusted once we get more data). Also store the transient errors into an azure table when still retrying, and a separate table when fail past the maximum retry attempts (tries a total of 4 times, one initial and 3 retries).

## Related issues
[AB#78067](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/78067)

## Testing
Updated a few unit tests and manually ran tests.

### CODE REVIEW

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/ExceptionStorage/ErrorType.cs [hunk 1]
```diff
@@ -11,9 +11,14 @@ namespace Microsoft.Health.DicomCast.Core.Features.ExceptionStorage
...
-        /// A transient error
+        /// A transient error that has reached the maximum number of retries
         /// </summary>
-        Transient,
+        TransientFailure,
+
+        /// <summary>
+        /// A transient error that will still be retried
+        /// </summary>
+        TransientRetry,
```
[the remark this answers is not in the document]
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/ExceptionStorage/ErrorType.cs:21]: Removed it, thanks for the catch

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Features/Storage/Models/Entities/RetryableEntity.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        /// </summary>
+        /// <param name="studyUid">StudyUID of the changefeed entry that failed </param>
+        /// <param name="seriesUid">SeriesUID of the changefeed entry that failed</param>
+        /// <param name="instanceUid">InstanceUID of the changefeed entry that failed</param>
+        /// <param name="changeFeedSequence">Changefeed sequence number that threw exception</param>
+        /// <param name="retryNum">Number of times changefeed entry has been retried</param>
+        /// <param name="ex">The exception that was thrown</param>
+        public RetryableEntity(string studyUid, string seriesUid, string instanceUid, long changeFeedSequence, int retryNum, Exception ex)
+        {
+            EnsureArg.IsNotNull(studyUid, nameof(studyUid));
```
[the remark this answers is not in the document]
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Features/Storage/Models/Entities/RetryableEntity.cs:25]: removed the setters, thanks

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Exceptions/DicomTagException.cs [hunk 3]
```diff
@@ -9,14 +9,9 @@ namespace Microsoft.Health.DicomCast.Core.Exceptions
 {
     public class DicomTagException : Exception
```
[the remark this answers is not in the document]
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Exceptions/DicomTagException.cs:10]: As discussed offline, leaving for now

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/ExceptionStorage/TableExceptionStore.cs [hunk 4]
```diff
@@ -29,5 +29,10 @@ public async Task WriteExceptionAsync(ChangeFeedEntry changeFeedEntry, Exception
         {
             await _store.StoreExceptionToTable(changeFeedEntry, exceptionToStore, errorType, cancellationToken);
         }
+
+        public async Task WriteRetryableExceptionAsync(ChangeFeedEntry changeFeedEntry, int retryNum, Exception exceptionToStore, CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/ExceptionStorage/TableExceptionStore.cs]: Updated as discussed

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs [hunk 5]
```diff
@@ -20,31 +20,49 @@ namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
...
+                .Handle<RetryableException>()
+                .WaitAndRetryAsync(
+                    maxRetryCount,
+                    retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
+                    async (exception, timeSpan, retryCount, context) =>
+                    {
+                        if (exception is RetryableException)
+                        {
+                            RetryableException ex = (RetryableException)exception;
+                            ChangeFeedEntry changeFeedEntry = ex.ChangeFeedEntry;
```
[the remark this answers is not in the document]
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs]: Thanks!

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Constants.cs [hunk 6]
```diff
@@ -15,7 +15,11 @@ internal static class Constants
 
         public const string DicomValidationTableName = "InvalidDicomTagExceptionTable";
 
+        public const string TransientFailureTableName = "TransientFailureExceptionTable";
+
+        public const string TransientRetryTableName = "TransientRetryExceptionTable";
+
         // List of all the tables that need to be initialized
-        public static readonly IEnumerable<string> AllTables = new List<string>() { FhirExceptionTableName, DicomValidationTableName, DicomExceptionTableName };
+        public static readonly IEnumerable<string> AllTables = new List<string>() { FhirExceptionTableName, DicomExceptionTableName, TransientFailureTableName, DicomValidationTableName, TransientRetryTableName };
```
[the remark this answers is not in the document]
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Constants.cs]: Updated

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Features/Storage/TableStore.cs [hunk 7]
```diff
@@ -78,5 +85,33 @@ public async Task StoreExceptionToTable(ChangeFeedEntry changeFeedEntry, Excepti
...
+            TableOperation operation = TableOperation.InsertOrMerge(entity);
+
+            try
+            {
+                await table.ExecuteAsync(operation, cancellationToken);
+                _logger.LogInformation("Retryable error when processsing changefeed entry: {ChangeFeedSequence} for DICOM instance with StudyUID: {StudyUID}, SeriesUID: {SeriesUID}, InstanceUID: {InstanceUID}. Tried {retryNum} time(s). Stored into table: {Table} in table storage.", changeFeedSequence, studyUid, seriesUid, instanceUid, retryNum, tableName);
+            }
+            catch
+            {
+                _logger.LogInformation("Retryable error when processsing changefeed entry: {ChangeFeedSequence} for DICOM instance with StudyUID: {StudyUID}, SeriesUID: {SeriesUID}, InstanceUID: {InstanceUID}. Tried {retryNum} time(s). Failed to store to table storage.", changeFeedSequence, studyUid, seriesUid, instanceUid, retryNum);
```
[the remark this answers is not in the document]
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Features/Storage/TableStore.cs:108]: The error thrown here will be logged further up the stack

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs [hunk 8]
```diff
@@ -20,35 +20,61 @@ namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
...
+                .Or<TaskCanceledException>()
+                .WaitAndRetryAsync(
+                    maxRetryCount,
+                    retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
+                    (exception, timeSpan, retryCount, context) =>
+                    {
+                        CancellationToken token = (CancellationToken)context[nameof(CancellationToken)];
+                        if (exception is TaskCanceledException && token.IsCancellationRequested)
+                        {
+                            return Task.CompletedTask;
```
Comment (wsugarman) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs:59]: Are you looking to retry if the `Task` was canceled? I see we're handling all `TaskCanceledException` instances, but presumably if we canceled we wouldn't want to retry, right?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs [hunk 9]
```diff
@@ -20,35 +20,61 @@ namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
     public class RetryableFhirTransactionPipeline : IFhirTransactionPipeline
     {
         private const int MaxRetryCount = 3;
-        private static readonly TimeSpan SleepDuration = TimeSpan.FromMilliseconds(100);
+
+        // private static readonly TimeSpan SleepDuration = TimeSpan.FromMilliseconds(100);
```
Comment (wsugarman) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs:24]: nit: are you going to delete this commented-out line?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Features/Storage/TableExceptionStore.cs [hunk 10]
```diff
@@ -78,5 +81,33 @@ public async Task StoreExceptionToTable(ChangeFeedEntry changeFeedEntry, Excepti
                 throw;
             }
         }
+
+        /// <inheritdoc/>
+        public async Task WriteRetryableExceptionAsync(ChangeFeedEntry changeFeedEntry, int retryNum, Exception exceptionToStore, CancellationToken cancellationToken)
```
Comment (wsugarman) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Features/Storage/TableExceptionStore.cs:86]: These methods are very similar. Is there a way to use a helper function to share some functionality?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/ChangeFeedProcessor.cs [hunk 11]
```diff
@@ -102,14 +103,18 @@ public async Task ProcessAsync(TimeSpan pollIntervalDuringCatchup, CancellationT
                             {
                                 errorType = ErrorType.DicomError;
                             }
+                            else if (ex is RetryableException || (ex is TaskCanceledException && !cancellationToken.IsCancellationRequested))
+                            {
+                                errorType = ErrorType.TransientFailure;
+                            }
 
                             await _exceptionStore.WriteExceptionAsync(
                             changeFeedEntry,
```
Comment (wsugarman) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/ChangeFeedProcessor.cs:112]: nit: missing indent for the arguments

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Features/Storage/TableExceptionStore.cs (hunk 10 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.TableStorage/Features/Storage/TableExceptionStore.cs:86]: As discussed off line will leave as is due to not enough overlap of functionality.

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs (hunk 8 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs]: Thanks for the catch, I updated it

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs [hunk 12]
```diff
@@ -20,35 +20,52 @@ namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
...
+            _exceptionStore = exceptionStore;
             _retryPolicy = Policy
-                .Handle<ResourceConflictException>()
-                .Or<ServerTooBusyException>()
-                .WaitAndRetryAsync(retryCount, sleepDurationProvider);
+                .Handle<RetryableException>()
+                .Or<TaskCanceledException>()
+                .WaitAndRetryAsync(
+                    maxRetryCount,
+                    retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
```
Comment (KnicKnic) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs:51]: we should probably have a max on the exponential. doubling every retry can get quite long quite quickly

Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs:51]: Do you mean a max number of times to retry? Currently it is set for 3. Do you think it should be changed?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs [hunk 13]
```diff
@@ -20,35 +20,52 @@ namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
...
+            EnsureArg.IsNotNull(exceptionStore, nameof(exceptionStore));
 
             _fhirTransactionPipeline = fhirTransactionPipeline;
+            _exceptionStore = exceptionStore;
             _retryPolicy = Policy
-                .Handle<ResourceConflictException>()
-                .Or<ServerTooBusyException>()
-                .WaitAndRetryAsync(retryCount, sleepDurationProvider);
+                .Handle<RetryableException>()
+                .Or<TaskCanceledException>()
```
Comment (wsugarman) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs:48]: I still think you want to change this `Or` condition to ignore cases where the user actually canceled the token. You don't want to retry those cases! However, the `HttpClient` timeout vs cancellation logic has changed over the years (especially in .NET Core). See here for more details: https://devblogs.microsoft.com/dotnet/net-5-new-networking-improvements/

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs [hunk 14]
```diff
@@ -20,35 +20,70 @@ namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
...
+            context[nameof(ChangeFeedEntry)] = changeFeedEntry;
+
+            return _retryPolicy.ExecuteAsync(
+                (ctx, tkn) =>
+                {
+                    try
+                    {
+                       return _fhirTransactionPipeline.ProcessAsync(changeFeedEntry, cancellationToken);
+                    }
+                    catch (TaskCanceledException ex)
```
Comment (wsugarman) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs:75]: nit: can use `catch (TaskExceptionException ex) when (!cancellationToken.IsCancellationRequested)`

**Review by wsugarman — APPROVED** [review]

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs (hunk 13 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs]: Updated in different manner as discussed

Comment (wsugarman) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs]: You don't need this anymore, right? You only need to catch `RetryableException`?

**Pull request merged**